File: BinarySearchTree/src/BinarySearchTree.java

public class BinarySearchTree {
	
	private class Node {
		public int value;
		public Node left;
		public Node right;
		
		public Node(int v) {
			value = v;
		}
	}
	
	private Node head;
	
	public boolean search(int v) {
		if (search(v, head) == null)
			return false;
		else
			return true;
	}
	
	private Node search(int v, Node head) {
		if (head == null || head.value == v)
			return head;
		else if (v < head.value)
			return search(v, head.left);
		else
			return search(v, head.right);
	}
	
	public void insert(int value) {
		if (head == null)
			head = new Node(value);
		else
			insert(value, head);
	}
	
	private void insert(int v, Node at) {
		if (v < at.value) {
			if (at.left == null)
				at.left = new Node(v);
			else
				insert(v, at.left);
		} else {
			if (at.right == null)
				at.right = new Node(v);
			else
				insert(v, at.right);
		}
	}
	
	public void delete(int value) {
		delete(value, head);
	}
	
	private void delete(int v, Node at) {
		if (at == null)
			return;
	}
	
	public void print() {
		print(head);
	}
	
	private void print(Node start) {
		if (start == null)
			return;
		print(start.left);
		System.out.println(start.value);
		print(start.right);
	}
}

File: BinarySearchTree_C/binary_search_tree.cpp
#include "binary_search_tree.hpp"

BinarySearchTree::BinarySearchTree()
{
	root = NULL;
};

bool BinarySearchTree::isEmpty()
{
	return root == NULL;
};

void BinarySearchTree::print()
{
	print(root);
}

void BinarySearchTree::print(node* at)
{
	if (at != NULL) 
	{	
		print(at->left);
		std::cout << at->value << std::endl;
		print(at->right);
	}
}

void BinarySearchTree::insertElement(int value)
{
	insertElement(value, root);
};

void BinarySearchTree::insertElement(int v, node* &at)
{
	if (at == NULL) 
	{
		at = new node(v);
	} 
	else
	{
		if (v > at->value) 
			insertElement(v, at->right);
		else
			insertElement(v, at->left);
	} 
}

bool BinarySearchTree::searchElement(int value)
{
	return searchElement(value, root);
};

bool BinarySearchTree::searchElement(int v, node* at)
{
	if (at == NULL) 
	{
		return false;
	} 
	else if (at->value == v) 
	{
		return true;
	}
	else
	{
		if (v > at->value) 
			return searchElement(v, at->right);
		else
			return searchElement(v, at->left);
	} 
}
